src/types.py

```diff
@@ -599,164 +846,106 @@ def unify_with_args(self, concrete: "FunctionSignature") -> bool:
-        if array_decay:
-            return Type.ptr(inner_type)
-        size = inner_type.get_size_bits()
-        if size is not None and dim is not None:
-            size *= dim
-        else:
-            size = None
-        return Type._ctype(real_ctype, typemap, size=size)
-    if isinstance(real_ctype, ca.PtrDecl):
-        return Type.ptr(type_from_ctype(real_ctype.type, typemap, array_decay=False))
-    if isinstance(real_ctype, ca.FuncDecl):
-        fn = parse_function(real_ctype)
-        assert fn is not None
-        fn_sig = FunctionSignature(
-            return_type=Type.void(),
-            is_variadic=fn.is_variadic,
-        )
-        if fn.ret_type is not None:
-            fn_sig.return_type = type_from_ctype(fn.ret_type, typemap)
-        if fn.params is not None:
-            fn_sig.params = [
-                FunctionParam(
-                    name=param.name or "",
-                    type=type_from_ctype(param.type, typemap),
-                )
-                for param in fn.params
-            ]
-            fn_sig.params_known = True
-        return Type.function(fn_sig)
-    if isinstance(real_ctype, ca.TypeDecl):
-        if isinstance(real_ctype.type, (ca.Struct, ca.Union)):
-            struct = parse_struct(real_ctype.type, typemap)
-            return Type._ctype(struct.type, typemap, size=struct.size * 8)
-        names = (
-            ["int"] if isinstance(real_ctype.type, ca.Enum) else real_ctype.type.names
+@dataclass(eq=False)
+class StructDeclaration:
+    """Representation of a C struct or union"""
+
+    @dataclass(eq=False)
+    class StructField:
+        type: Type
+        offset_bits: int
+        name: str
+
+    size_bits: int
+    align_bits: int
+    tag_name: Optional[str]
+    typedef_name: Optional[str]
+    fields: List[StructField]  # sorted by `.offset_bits`
+    has_bitfields: bool
+    is_union: bool
+
+    def unify(
+        self,
+        other: "StructDeclaration",
+        *,
+        seen: Optional[Set["TypeData"]] = None,
+    ) -> bool:
+        # NB: Currently, the only structs that exist are defined from ctypes in the typemap,
+        # so for now we can use reference equality to check if two structs are compatible.
+        return self is other
+
+    def fields_at_offset(self, offset_bits: int) -> List[StructField]:
+        """Return the list of StructFields which contain the given offset (in bits)"""
+        fields = []
+        for field in self.fields:
+            # We assume fields are sorted by `offset_bits`, ascending
+            if field.offset_bits > offset_bits:
+                break
+            field_size_bits = field.type.get_size_bits()
+            assert field_size_bits is not None
+            if field.offset_bits + field_size_bits < offset_bits:
+                continue
+            fields.append(field)
+        return fields
+
+    @staticmethod
+    def from_ctype(ctype: Union[ca.Struct, ca.Union], universe: TypeUniverse) -> Type:
```

Comment:
Yeah, I admit that is an awkward part of this API.

I ended up in this situation to make it easier to construct self-referential structs (like linked lists).
The code ends up inserting the `Type` object that *will* be returned into the universe before the `StructDeclaration` fields are fully populated. This makes self-reference easy, since it makes the struct `Type` available in the universe while parsing the fields.

I'll try to revisit this.

Reply:
I see. Would be useful to have a comment about this if revisiting doesn't work out. (I feel like the same trick should work by inserting the StructDeclaration before it's fully constructed, but I'm probably missing some reason why it gets involved)